SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -39,180 +39,184 @@ void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
 	{
-		return Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses);
+		return FSpatialGDKSpanId(false);
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	FSpatialGDKSpanId TraceSpanId(true);
+	FMemory::Memcpy(TraceSpanId.GetData(), UserSpanId.Data.GetData(), TRACE_SPAN_ID_SIZE_BYTES);
+	return TraceSpanId;
 }
 
-void SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const TOptional<Trace_SpanId>& OptionalSpanId)
+FSpatialGDKSpanId SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const Trace_SpanIdType* Causes /* = nullptr*/, int32 NumCauses /* = 0*/)
 {
-	if (!IsEnabled())
+	if (Causes == nullptr && NumCauses > 0)
 	{
-		return;
+		return FSpatialGDKSpanId(false);
 	}
 
 	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
-	const ANSICHAR* Message = MessageSrc.Get();
+	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type.ToString());
 
-	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type);
-	const ANSICHAR* Type = TypeSrc.Get();
+	Trace_Event Event;
+	Event.span_id = nullptr;
+	Event.message = MessageSrc.Get();
+	Event.type = TypeSrc.Get();
+	Event.unix_timestamp_millis = 0;
+	Event.data = nullptr;
 
-	Trace_Event TraceEvent{ OptionalSpanId.GetValue(), /* unix_timestamp_millis: ignored */ 0, Message, Type, nullptr };
-	if (!Trace_EventTracer_ShouldSampleEvent(EventTracer, &TraceEvent))
+	Trace_SamplingResult SpanSamplingResult = Trace_EventTracer_ShouldSampleSpan(EventTracer, Causes, NumCauses, &Event);
+	if (SpanSamplingResult.decision == Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE)
 	{
-		return;
+		return FSpatialGDKSpanId(true);
 	}
 
-	Trace_EventData* EventData = Trace_EventData_Create();
+	FSpatialGDKSpanId TraceSpanId(true);
+	Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses, &Event, TraceSpanId.GetData());
+	Event.span_id = TraceSpanId.GetData();
 
-	for (const auto& Pair : SpatialTraceEvent.Data)
+	Trace_SamplingResult EventSamplingResult = Trace_EventTracer_ShouldSampleEvent(EventTracer, &Event);
+	switch (EventSamplingResult.decision)
+	{
+	case Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE:
 	{
-		auto KeySrc = StringCast<ANSICHAR>(*Pair.Key);
-		const ANSICHAR* Key = KeySrc.Get();
-		auto ValueSrc = StringCast<ANSICHAR>(*Pair.Value);
-		const ANSICHAR* Value = ValueSrc.Get();
-		Trace_EventData_AddStringFields(EventData, 1, &Key, &Value);
+		return TraceSpanId;
 	}
-
-	// Frame counter
+	case Trace_SamplingDecision::TRACE_SHOULD_SAMPLE_WITHOUT_DATA:
 	{
-		const char* FrameCountStr = "FrameNum";
-		char TmpBuffer[64];
-		FCStringAnsi::Sprintf(TmpBuffer, "%" PRIu64, GFrameCounter);
-		const char* TmpBufferPtr = TmpBuffer;
-		Trace_EventData_AddStringFields(EventData, 1, &FrameCountStr, &TmpBufferPtr);
+		Trace_EventTracer_AddEvent(EventTracer, &Event);
+		return TraceSpanId;
 	}
+	case Trace_SamplingDecision::TRACE_SHOULD_SAMPLE:
+	{
+		Trace_EventData* EventData = Trace_EventData_Create();
```

Comment:
Nit: Can do this outside a switch statement - the two return cases can be above and a single AddEvent can exist with just a branch for TRACE_SHOULD_SAMPLE - think it's safe to assume the decision is within the valid return types

Reply:
Just wanted to point out that we do something in the TRACE_SHOULD_SAMPLE_WITHOUT_DATA slightly different to the TRACE_SHOULD_NOT_SAMPLE case. Sorry if you realized that already.